### PR TITLE
add PositionComponent children getter

### PR DESCRIPTION
**Pull request by unger1984**
# Description

Add getter for private property _children. This allows parents to control they children. Child not know about the global position, only parent this known, and for this, the expanding parent class must control his children.

## Type of change

New feature (non-breaking change which adds functionality)

# Checklist:

If something is unclear, please submit the PR anyways and ask about what you thought was unclear.

- [x] This branch is based on `v1.0.0`
- [ ] This PR is targeted to merge into `v1.0.0` (not `master` or `develop`)
- [x] I have added an entry under `[next]` in `CHANGELOG.md`
- [x] I have formatted my code with `flutter format`
- [ ] I have made corresponding changes to the documentation
- [ ] I have added examples for new features in `doc/examples`
- [ ] The continuous integration (CI) is passing

### CODE REVIEW

**Comment by spydon** [comment]
I have a PR coming in soon with a new suggested way of handling children and propagation, but we can keep this open meanwhile if we decide to reject that one.

**Comment by unger1984** [comment]
> I have a PR coming in soon with a new suggested way of handling children and propagation, but we can keep this open meanwhile if we decide to reject that one.

I understand that changing the properties of children from a parent can be dangerous. But I just don’t know any other way to call the child's methods from the parent other than propagateToChildren. But propagateToChildren is generally not suitable for many cases when it is necessary to affect some of the children and not all.

We can mark the proposed method as dangerous in the documentation.

**Comment by unger1984** [comment]
Use case:

```dart
 void onDragStart(Offset position) {
    getChildren.forEach((child) {
      if (child is PositionComponent) {
        final childRect =
            Vector2(x + child.x, y + child.y).toPositionedRect(child.size);
        if (childRect.contains(position)) {
          if (child is Draggable) {
            child.onDragStart(position);
          } else if (child is HasDraggable) {
            child.onDragStart(position);
          }
        }
      }
    });
  }
```

**Comment by unger1984** [comment]
Full code of example:

```dart
import 'dart:ui';

import 'package:flame/components/component.dart';
import 'package:flame/components/position_component.dart';
import 'package:flame/extensions/vector2.dart';
import 'package:flame/flame.dart';
import 'package:flame/game.dart';
import 'package:flame/gestures.dart';
import 'package:flame/text_config.dart';
import 'package:flutter/material.dart' hide Animation, Image, Draggable;

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await Flame.util.fullScreen();

  Vector2 size = await Flame.util.initialDimensions();
  MyGame game = MyGame(size);
  runApp(game.widget);
}

class Child extends PositionComponent with Draggable {
  Child() {
    debugMode = true;
    position = Vector2(20, 40);
    size = Vector2(160, 100);
  }

  @override
  void render(Canvas canvas) {
    super.render(canvas);
    TextConfig(color: debugColor, fontSize: 12).render(canvas,
        'Child with Draggable\n Mast drag only they touch!', Vector2(10, 10));
  }
}

class Test extends PositionComponent with HasDraggable {
  Test() {
    debugMode = true;
    position = Vector2.all(100);
    size = Vector2(250, 200);

    addChild(null, Child());
  }

  @override
  void render(Canvas canvas) {
    super.render(canvas);
    TextConfig(color: debugColor, fontSize: 12)
        .render(canvas, 'fixed HasDraggable', Vector2(10, 10));
  }
}

class TestWindow extends GameWindow with Draggable {
  TestWindow(Vector2 s, Vector2 p) : super(s, p);

  @override
  void render(Canvas canvas) {
    super.render(canvas);
    TextConfig(color: debugColor, fontSize: 12)
        .render(canvas, 'Draggable', Vector2(10, 10));
  }
}

class TestDesktop extends GameDesktop {
  GameWindow window;

  TestDesktop(Vector2 gameSize) : super(gameSize) {
    window = TestWindow(Vector2(200, 100), Vector2(20, 350));
    add(window);
    add(Test());
  }

  @override
  void render(Canvas canvas) {
    super.render(canvas);
    TextConfig(color: debugColor, fontSize: 12)
        .render(canvas, 'Global Parent HasDraggable', Vector2(10, 10));
  }
}

class MyGame extends Game with MultipleDetector {
  Vector2 gameSize;
  GameDesktop desktop;

  MyGame(this.gameSize) {
    desktop = TestDesktop(gameSize);
  }

  @override
  Color backgroundColor() => const Color(0xFF38607C);

  @override
  void render(Canvas canvas) {
    if (desktop != null) desktop.render(canvas);
  }

  @override
  void update(double t) {
    if (desktop != null) desktop.update(t);
  }

  @override
  void onResize(Vector2 size) {
    if (desktop != null) desktop.onGameResize(size);
  }

  @override
  void onDragUpdate(Offset details) {
    if (desktop != null) desktop.onDragUpdate(details);
  }

  @override
  void onDragStart(Offset position) {
    if (desktop != null) desktop.onDragStart(position);
  }

  @override
  void onDragEnd() {
    if (desktop != null) desktop.onDragEnd();
  }
}

enum ScaleType { drag, scale, none, unknown }

mixin MultipleDetector on Game implements ScaleDetector {
  ScaleType _type = ScaleType.none;
  Offset _sacleStart;
  double _oldScale = 0.0;

  @override
  void onScaleStart(ScaleStartDetails details) {
    _type = ScaleType.unknown;
    _sacleStart = details.focalPoint;
  }

  @override
  void onScaleUpdate(ScaleUpdateDetails details) {
    // Если еще не известно что за событие
    if (_type == ScaleType.unknown) {
      if (details.scale == 1) {
        _type = ScaleType.drag;
        onDragStart(_sacleStart);
      } else {
        _type = ScaleType.scale;
      }
    }

    switch (_type) {
      case ScaleType.scale:
        if (details.scale != _oldScale) {
          onScaleChange(details.scale);
          _oldScale = details.scale;
        }
        break;
      case ScaleType.drag:
        final nposition = details.focalPoint;
        final dx = nposition.dx - _sacleStart.dx;
        final dy = nposition.dy - _sacleStart.dy;
        onDragUpdate(Offset(dx, dy));
        break;
      case ScaleType.none:
      case ScaleType.unknown:
        break;
    }
  }

  @override
  void onScaleEnd(ScaleEndDetails details) {
    if (_type == ScaleType.drag) {
      onDragEnd();
    }
    _type = ScaleType.none;
  }

  void onScaleChange(double scale) {}
  void onDragUpdate(Offset details) {}
  void onDragStart(Offset position) {}
  void onDragEnd() {}
}

mixin HasDraggable on PositionComponent {
  void onDragStart(Offset position) {
    getChildren.forEach((child) {
      if (child is PositionComponent) {
        final childRect =
            Vector2(x + child.x, y + child.y).toPositionedRect(child.size);
        if (childRect.contains(position)) {
          if (child is Draggable) {
            child.onDragStart(position);
          } else if (child is HasDraggable) {
            child.onDragStart(position);
          }
        }
      }
    });
  }

  void onDragUpdate(Offset details) {
    getChildren.forEach((child) {
      if (child is PositionComponent) {
        if (child is Draggable) {
          child.onDragUpdate(details);
        } else if (child is HasDraggable) {
          child.onDragUpdate(details);
        }
      }
    });
  }

  @override
  void onDragEnd() {
    getChildren.forEach((child) {
      if (child is PositionComponent) {
        if (child is Draggable) {
          child.onDragEnd();
        } else if (child is HasDraggable) {
          child.onDragEnd();
        }
      }
    });
  }
}

mixin Draggable on PositionComponent {
  Vector2 _oldPosition = Vector2.zero();
  bool isDragged = false;
  Vector2 dragLimit;

  void onDragStart(Offset position) {
    _oldPosition = Vector2(x, y);
    isDragged = true;
  }

  void onDragUpdate(Offset details) {
    if (isDragged) {
      final old = Vector2(x, y);
      x = _oldPosition.x + details.dx;
      y = _oldPosition.y + details.dy;
      if (dragLimit != null) {
        if (x > 0) x = 0;
        if (y > 0) y = 0;
        if (x + size.x < dragLimit.x) x = dragLimit.x - size.x;
        if (y + size.y < dragLimit.y) y = dragLimit.y - size.y;
      }
    }
  }

  void onDragEnd() {
    _oldPosition = Vector2(x, y);
    isDragged = false;
  }
}

class GameDesktop extends PositionComponent with HasDraggable {
  GameDesktop(Vector2 gameSize) {
    x = 0;
    y = 0;
    size = gameSize;
    debugMode = true;
  }

  void add(Component c) {
    addChild(null, c);
  }

  @override
  Color get debugColor => const Color(0xFF00FF00);

  @override
  void onGameResize(Vector2 gameSize) {
    size = gameSize;
    super.onGameResize(gameSize);
  }
}

class GameWindow extends PositionComponent {
  GameWindow(Vector2 s, Vector2 p) {
    size = s;
    position = p;
    debugMode = true;
  }

  @override
  Color get debugColor => const Color(0xFF0000FF);
}
```

**Comment by spydon** [comment]
I think I get your use-case, the `children` list won't be private in my PR, so you'll have direct access to it.

**Comment by unger1984** [comment]
> I think I get your use-case, the `children` list won't be private in my PR, so you'll have direct access to it.

When will your PR be ready?

File: lib/components/position_component.dart [hunk 1]
```diff
@@ -212,4 +212,6 @@ abstract class PositionComponent extends Component {
   void clearChildren() {
     _children.clear();
   }
+
+  OrderedSet<Component> get getChildren => _children;
```
Comment (luanpotter) [inline lib/components/position_component.dart:216]: I would:
1. call it `children`
2. copy

now, do you have an example of what you are trying to do? I wonder if there is something else we can expose to make it even easier (so the children knows what they need to know)

Comment (unger1984) [reply lib/components/position_component.dart:216]: example added to PR comments

**Comment by spydon** [comment]
I'll close this one and we can figure out a way forward in #548. :)

**Pull request closed without merging**